// @mui/x-data-grid-premium/components/columnMenu/menuItems/GridColumnMenuChartsItem.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GridColumnMenuChartsItem = GridColumnMenuChartsItem;
var React = _interopRequireWildcard(require("react"));
var _xDataGridPro = require("@mui/x-data-grid-pro");
var _gridChartsIntegrationSelectors = require("../../../hooks/features/chartsIntegration/gridChartsIntegrationSelectors");
var _useGridRootProps = require("../../../hooks/utils/useGridRootProps");
var _useGridApiContext = require("../../../hooks/utils/useGridApiContext");
var _jsxRuntime = require("react/jsx-runtime");
function GridColumnMenuChartsItem(props) {
  const {
    onClick
  } = props;
  const rootProps = (0, _useGridRootProps.useGridRootProps)();
  const apiRef = (0, _useGridApiContext.useGridApiContext)();
  const isChartsPanelOpen = (0, _xDataGridPro.useGridSelector)(apiRef, _gridChartsIntegrationSelectors.gridChartsPanelOpenSelector);
  const openChartsSettings = event => {
    onClick(event);
    apiRef.current.setChartsPanelOpen(true);
  };
  if (!rootProps.experimentalFeatures?.charts || !rootProps.chartsIntegration) {
    return null;
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(rootProps.slots.baseMenuItem, {
    onClick: openChartsSettings,
    iconStart: /*#__PURE__*/(0, _jsxRuntime.jsx)(rootProps.slots.chartsIcon, {
      fontSize: "small"
    }),
    disabled: isChartsPanelOpen,
    children: apiRef.current.getLocaleText('columnMenuManageCharts')
  });
}
